clash-lib/src/Clash/Primitives/Xilinx/ClockGen.hs

```diff
@@ -0,0 +1,126 @@
+{-|
+  Copyright   :  (C) 2023, QBayLogic B.V.
+                     2023, Google Inc.
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Blackbox template functions for
+  Clash.Xilinx.ClockGen.{clockWizard,clockWizardDifferential}
+-}
+
+{-# LANGUAGE QuasiQuotes #-}
+
+module Clash.Primitives.Xilinx.ClockGen where
+
+import Clash.Backend                      (Backend)
+import Clash.Netlist.BlackBox.Util        (exprToString)
+import Clash.Netlist.Types
+import Clash.Netlist.Util                 (stripVoid)
+
+import Control.Monad.State                (State)
+import qualified Data.String.Interpolate  as I
+import Data.Text.Prettyprint.Doc.Extra    (Doc)
+
+
+clockWizardTclTF :: TemplateFunction
+clockWizardTclTF =
+  TemplateFunction used valid clockWizardTclTemplate
+ where
+  used = [0..4]
+  valid bbCtx
```

Comment:
I missed this bit in my review. We've discussed the `used` in the office. Furthermore, this `valid` is incomplete, as it doesn't check `kdIn` and `kdOut` are actually `KnownDomain` constraints. But as far as I'm concerned, you make `valid` a `const True` and let the `TclTemplate` handle printing the error. In fact, I prefer the error message that produces :-).